### PR TITLE
Add kernel.org's `.clang-format` for editor-agnostic C formatting hints

### PR DESCRIPTION
**Pull request by bleggett**
`clang-format` is an editor-agnostic means of describing C formatting prefs for code editors.

Since upstream kernel uses this to define an explicit contract for C formatting, and our
BPF code follows the kernel style, this just brings in a copy of the kernel `.clang-format` dotfile.

This will get picked up by any editor that supports it (VSCode w/ C extension, emacs with `lsp`, vims, etc).

This will declare a specific contract for contributors around C style, as well as help this project explicitly declare any deviations from that style.

This is a straight copy of the [kernel.org file](https://github.com/torvalds/linux/blob/master/.clang-format), with a kernel-specific macro allowlist removed.

### CODE REVIEW

**Comment by qmonnet** [comment]
Thanks for this!

I know we've discussed using `clang-format` before, but we never really reached a consensus. Sadly, Slack history is not persistent and these discussions are lost.

For my part, I'd be happy to defer formatting to `clang-format`, but last time I checked it would format some elements in a strange way, even with the kernel's config. This is maybe just a matter of fine-tuning the correct options, but I haven't spent time on that.

> kernel uses this to define an explicit contract for C formatting, and our BPF code follows the kernel style

Out of curiosity, do you have a reference for this? My understanding was that the config file was here to help, but the style is not enforced and I don't think it's even considered as an absolute reference, more like a best-effort config? If anything, `checkpatch.pl` is the one that runs in BPF or Netdev CIs.

> This will get picked up by any editor that supports it (VSCode w/ C extension, emacs with `lsp`, vims, etc).

What does it do? Does it format the new code accordingly, or does it reformat the entire file on save? The former sounds good, the latter not that much - the current formatting is too far from what `clang-format` do and that would mess up with commits?

> This will declare a specific contract for contributors around C style, as well as help this project explicitly declare any deviations from that style.

Yes, this intent is good.

> This is a straight copy of the [kernel.org file](https://github.com/torvalds/linux/blob/master/.clang-format), with a kernel-specific macro allowlist removed.

So a few more questions:

- Did you run it on the current sources and did you find anything weird-looking? On a quick look, I note at least the following:
  - When declaring a function, it always moves the return type to the same line as the function name, when we usually have it on a separate line if the type is long (which is often the case given that we've got plenty of `static __always_inline ...`).
  - It always remove spaces between `#` and preprocessor directives (`# ifdef`), but these spaces are sometimes used to visually “indent” nested directives.
  - It struggles with code indent for blocks that are under `#ifdef`s, I've spotted several locations where it would bring it all the way to the left instead of keeping the indent from the current context (Note: I realised afterwards that I used `clang-format` from LLVM 10, this might have been improved since? TBC).
  - Likely more things to say, but I haven't done an exhaustive pass.

- Related question: Is there a way to turn off `clang-format` locally? I know some linters support this. I understand you're not proposing to enforce the style in CI checks yet, but this will inevitably come up at some point if we start using `clang-format` on a regular basis.

- Tetragon uses `clang-format`, it might be worth looking at [the amendments they did to the kernel config file](https://github.com/cilium/tetragon/blob/main/.clang-format) to see if there's anything relevant to Cilium, as well.

TL;DR: Happy with the idea of using `clang-format` for Cilium, but I've got concerns with some elements that are not formatted correctly.

Cc @aditighag @christarazi who were involved in previous discussions, if I remember correctly.

**Comment by bleggett** [comment]
> For my part, I'd be happy to defer formatting to `clang-format`, but last time I checked it would format some elements in a strange way, even with the kernel's config. This is maybe just a matter of fine-tuning the correct options, but I haven't spent time on that.


> > kernel uses this to define an explicit contract for C formatting, and our BPF code follows the kernel style
> 
> Out of curiosity, do you have a reference for this? My understanding was that the config file was here to help, but the style is not enforced and I don't think it's even considered as an absolute reference, more like a best-effort config? If anything, `checkpatch.pl` is the one that runs in BPF or Netdev CIs.

`checkpatch` is the actual enforced contract (for the kernel and here) yes - but it helps to have a declarative contract that editors understand so you don't have to wait until you run up against `checkpatch`.

So I'm not proposing we replace `checkpatch` CI checks with clang-format (at least not until/unless kernel tree does) - better to have both. One is to make local dev easier, the other is a CI gate.

> > This will get picked up by any editor that supports it (VSCode w/ C extension, emacs with `lsp`, vims, etc).
> 
> What does it do? Does it format the new code accordingly, or does it reformat the entire file on save? The former sounds good, the latter not that much - the current state of Cilium's code formatting is too far from what `clang-format` does and that would mess up with commits?

It works like `editorconfig` and configures editor formatting rules. Whenever your editor is told to format C code in this repo or subfolders, if it is configured to use the [clang-format](https://clang.llvm.org/docs/ClangFormat.html) tool it will try to do so according to the rules in `.clang-format`. How and where you choose to apply formatting is a local thing, and for all the editors I know of, auto-reformatting is not the default behavior.

It would be opt-in/only apply if a manual reformat was triggered. I find it helpful since I already use LSP-based extensions and largely don't want to spend time on manual formatting. Someone would have to manually reformat every C file in the repo for this to apply globally, and that's certainly not the intent.

> So a few more questions:
> 
>     * Did you run it on the current sources and did you find anything weird-looking? 

Not on all files, but yeah I did notice some stylistic differences. Figured I'd raise this PR and see what people thought about it.

- If there's something that breaks `checkpatch`, that is something to fix - and anyway checkpatch will catch those corner cases which can be fixed manually, so it's exactly as safe as the existing flow. And anyway, if people have non-`clang-format` based formatting tools/extensions, they can keep using those and ignore this.

- If there's something that doesn't break `checkpatch` but is a stylistic change from current convention, we should either make `.clang-format` respect it via an explicit rule, or accept the default - either option is useful.


> On a quick look, I note at least the following:
>       
>       * When declaring a function, it always moves the return type to the same line as the function name, when we usually have it on a separate line if the type is long (which is often the case given that we've got plenty of `static __always_inline ...`).
>       * It always remove spaces between `#` and preprocessor directives (`# ifdef`), but these spaces are sometimes used to visually “indent” nested directives.
>       * It struggles with code indent for blocks that are under `#ifdef`s, I've spotted several locations where it would bring it all the way to the left instead of keeping the indent from the current context (Note: I realised afterwards that I used `clang-format` from LLVM 10, this might have improved since? TBC).
>       * Likely more things to say, but I haven't done an exhaustive pass.

TY, I'll look into these and see if they can be explicitly configured.

>     * Related question: Is there a way to turn off `clang-format` locally? I know some linters support this. I understand you're not proposing to enforce the style in CI checks yet, but this will inevitably come up at some point if we start using `clang-format` on a regular basis.

Turn off your editor's `clang-format` extension (or extension option) - or just locally delete `.clang-format` if you don't want it to use these rules. Again - pretty much the same as `editorconfig`.

>     * When this time comes and we finally consider it for CI, we'd need to sort out how it plays with the current checkpatch validation (how to solve conflicts between the two, and so on).

Yeah as mentioned above - don't wanna do that until kernel tree does the same. In the meantime this is purely a local dev aid to help avoid roundtrips against `checkpatch` failures.

>     * Tetragon uses `clang-format`, it might be worth looking at [the amendments they did to the kernel config file](https://github.com/cilium/tetragon/blob/main/.clang-format) to see if there's anything relevant to Cilium, as well.

Oh nice I missed that, will take a look.

> TL;DR: Happy with the idea of using `clang-format` for Cilium, but I've got concerns on how exactly we'd use it, and with some elements that are not formatted correctly.

Thanks for the feedback! I figured that would be the case and it'd be worth discussing. I'll see if I can address those concerns.

**Comment by qmonnet** [comment]
> `checkpatch` is the actual enforced contract (for the kernel and here) yes - but it helps to have a declarative contract that editors understand so you don't have to wait until you run up against `checkpatch`.
> 
> So I'm not proposing we replace `checkpatch` CI checks with clang-format (at least not until/unless kernel tree does) - better to have both. One is to make local dev easier, the other is a CI gate.

OK, agreed

> It works like `editorconfig` and configures editor formatting rules. Whenever your editor is told to format C code in this repo or subfolders, if it is configured to use the [clang-format](https://clang.llvm.org/docs/ClangFormat.html) tool it will try to do so according to the rules in `.clang-format`. How and where you choose to apply formatting is a local thing, and for all the editors I know of, auto-reformatting is not the default behavior.

Which is good, OK thanks for clarifying

> It would be opt-in/only apply if a manual reformat was triggered. I find it helpful since I already use LSP-based extensions and largely don't want to spend time on manual formatting. Someone would have to manually reformat every C file in the repo for this to apply globally, and that's certainly not the intent.

:+1: 

> [...] I did notice some stylistic differences. Figured I'd raise this PR and see what people thought about it.
> 
>  * If there's something that breaks `checkpatch`, that is something to fix - and anyway `checkpatch` will catch those corner cases which can be fixed manually, so it's exactly as safe as the existing flow. And  if people have non-`clang-format` based formatting tools/extensions, they can keep using those and ignore this.
> 
>  * If there's something that doesn't break `checkpatch` but is a stylistic change from current convention, we should either make `.clang-format` respect it via an explicit rule, or accept the default - either option is useful.

Agreed. I guess whether we prefer to keep the current style or adopt `clang-format`'s is something that we need to figure out on a case-by-case basis. The important thing for me at the moment is that by introducing the tool, it does not kinda _force_ us to adopt all its suggestions. There are a few locations, especially when doing complex macro declarations, where elements don't follow the usual formatting rules, and we likely want to keep it that way.

> TY, I'll look into these and see if they can be explicitly configured.

Yes please, this would be great!

> Turn off your editor's `clang-format` extension (or extension option) - or just locally delete `.clang-format` if you don't want it to use these rules. Again - pretty much the same as `editorconfig`.
> 
> This is purely opt-in locally for everybody, except for people that might have already been using a plugin that makes use of `clang-format` and was using some other set of defaults, and didn't realize it.

OK. I meant more about turning it off for a portion of code if we ever adopt the tool in CI, and can't encode the style we want with clang-format options. I see it should be doable if needed: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code:

```c
int formatted_code;
// clang-format off
    void    unformatted_code  ;
// clang-format on
void formatted_code_again;
```

> Thanks for the feedback! I figured that would be the case and it'd be worth discussing. I'll see if I can address those concerns.

Yes, we've discussed it before and although we never got a consensus, I think it would be a good thing overall to agree on a style definition and get the `.clang-format` file merged.

Please do take a look at the Tetragon changes, and at whether it seems doable to address the style changes mentioned above, if you have a moment. Would be great to have something as close to the existing style as possible. Then it will be good to go as far as I'm concerned, although I'll try to mention it at the next community meeting (Wednesday), for awareness and to make sure nobody has major concerns about it (I don't expect there will be).

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+FixNamespaceComments: false
+
+IncludeBlocks: Preserve
+IncludeCategories:
+  - Regex: '.*'
+    Priority: 1
+IncludeIsMainRegex: '(Test)?$'
+IndentCaseLabels: false
+IndentGotoLabels: false
+IndentPPDirectives: None
```
Comment (qmonnet) [inline .clang-format:84]: We could use this:

```suggestion
IndentPPDirectives: AfterHash
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#indentppdirectives

Although this indents directives everywhere, including where we're not currently using them. Meaning that if, for example, we remove one level of nested `#ifdef`s, we have to update the indent for inner directives.

If using the above we'd also need to set `PPIndentWidth`, because it seems to default to `IndentWidth` but we don't want that:

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#ppindentwidth

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+IndentWidth: 8
```
Comment (qmonnet) [inline .clang-format:85]:
```suggestion
IndentWidth: 8
InsertNewlineAtEOF: true
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#insertnewlineateof

File: .clang-format [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+# Taken from git's rules
+PenaltyBreakAssignment: 10
+PenaltyBreakBeforeFirstCallParameter: 30
+PenaltyBreakComment: 10
+PenaltyBreakFirstLessLess: 0
+PenaltyBreakString: 10
+PenaltyExcessCharacter: 100
+PenaltyReturnTypeOnItsOwnLine: 60
+
+PointerAlignment: Right
```
Comment (qmonnet) [inline .clang-format:113]:
```suggestion
PointerAlignment: Right
QualifierAlignment: Left
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#qualifieralignment

Note: There's an option to specify the list for sorting the different qualifiers, but I haven't been able to get it to work on clang-format-17.

File: .clang-format [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+ReflowComments: false
+SortIncludes: false
+SortUsingDeclarations: false
+SpaceAfterCStyleCast: false
+SpaceAfterTemplateKeyword: true
+SpaceBeforeAssignmentOperators: true
+SpaceBeforeCtorInitializerColon: true
+SpaceBeforeInheritanceColon: true
+SpaceBeforeParens: ControlStatements
+SpaceBeforeRangeBasedForLoopColon: true
```
Comment (qmonnet) [inline .clang-format:126]: Seems like we're leaving no space in `{}` most of the time, we could as well enforce it.

```suggestion
SpaceBeforeRangeBasedForLoopColon: true
SpaceInEmptyBlock: true
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#spaceinemptyblock

File: .clang-format [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+#
+# Obtained from: https://github.com/torvalds/linux/blob/master/.clang-format
+---
+AccessModifierOffset: -4
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
+AlignEscapedNewlines: Left
+AlignOperands: true
+AlignTrailingComments: false
```
Comment (qmonnet) [inline .clang-format]: Looks like we might to change this one, we're aligning comments at multiple locations

```suggestion
AlignTrailingComments: Always
```

File: .clang-format [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+#   Documentation/process/clang-format.rst
+#   https://clang.llvm.org/docs/ClangFormat.html
+#   https://clang.llvm.org/docs/ClangFormatStyleOptions.html
+#
+# Obtained from: https://github.com/torvalds/linux/blob/master/.clang-format
+---
+AccessModifierOffset: -4
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
```
Comment (qmonnet) [inline .clang-format:21]: Note: We would like to enable these inside of struct/enum/macro declarations and definitions, but it looks like it's not possible :disappointed:. Looks like it's either aligned or not, for everywhere.

Comment (qmonnet) [inline .clang-format:21]: We can at least enable for macros, though:

```suggestion
AlignConsecutiveDeclarations: false
AlignConsecutiveMacros:
  Enabled: true
  AcrossEmptyLines: true
  AcrossComments: true
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignconsecutivemacros

File: .clang-format [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+AllowShortCaseLabelsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: None
+AllowShortIfStatementsOnASingleLine: false
+AllowShortLoopsOnASingleLine: false
+AlwaysBreakAfterDefinitionReturnType: None
+AlwaysBreakAfterReturnType: None
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: false
+BinPackArguments: true
+BinPackParameters: true
```
Comment (qmonnet) [inline .clang-format:40]:
```suggestion
BinPackParameters: true
BitFieldColonSpacing: None
```

`clang-format` also seems to move bitfields components onto a single line, but I've not found how to disable that :disappointed:.

We're also getting spaces before colons for labels, but I'm not sure how to disable that, either.

File: .clang-format [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+ObjCSpaceBeforeProtocolList: true
+
+# Taken from git's rules
+PenaltyBreakAssignment: 10
+PenaltyBreakBeforeFirstCallParameter: 30
+PenaltyBreakComment: 10
+PenaltyBreakFirstLessLess: 0
+PenaltyBreakString: 10
+PenaltyExcessCharacter: 100
+PenaltyReturnTypeOnItsOwnLine: 60
```
Comment (qmonnet) [inline .clang-format]: Looks like we need to experiment with this value, but I haven't been able to understand how it works, yet.

**Review by qmonnet — COMMENTED** [review]
Various comments and observations. I experimented a little, not something exhaustive.

(Additions or updates to the values we got from the kernel's file, if any, should likely go into a separate commit.)

File: .clang-format (hunk 3 above)
Comment (bleggett) [reply .clang-format:113]: I'm inclined to leave this one as `Leave` since it apparently can break things/change intent if set to anything else.

File: .clang-format (hunk 4 above)
Comment (bleggett) [reply .clang-format:126]: `checkpatch` has no opinions about this AFAICT, and the current codebase default seems to be `SpaceInEmptyBlock: false`.

When `checkpatch` doesn't have an opinion, it's probably best to go with a setting that reflects the existing style already in the codebase (for now, anyway) - I can't find any existing instances of `{ }` on a cursory look.

File: .clang-format (hunk 6 above)
Comment (bleggett) [reply .clang-format:21]: For now, suggest leaving it as `false` - yep.

**Comment by bleggett** [comment]
> Various comments and observations. I experimented a little, not something exhaustive.
> 
> (Additions or updates to the values we got from the kernel's file, if any, should likely go into a separate commit.)

My fedora defaults to clang-format 15 - some of the options you added are in 16 only.

I'm inclined to track (current version - 2) just to keep things simple.

File: .clang-format (hunk 7 above)
Comment (bleggett) [reply .clang-format:40]: Can you point me at a file where you see this?

File: .clang-format (hunk 4 above)
Comment (qmonnet) [reply .clang-format:126]: Apologies, I meant to set it to `false`, but messed up when writing the comment.

File: .clang-format (hunk 7 above)
Comment (qmonnet) [reply .clang-format:40]: - For bitfields: look for `[^ ]:[^ ]`, first example I see is in `struct policy_entry` in `bpf/lib/common.h`.

- For the space between label and colon: `skip_egress_gateway` in `bpf/bpf_lxc.c` is the first example I could find.

**Comment by qmonnet** [comment]
> My fedora defaults to clang-format 15 - some of the options you added are in 16 only.
> 
> I'm inclined to track (current version - 2) just to keep things simple - that seems to track with Ubuntu (and presumably other) distros as well.

Makes sense. I didn't pay too much attention to the required version when looking at these options, so indeed it makes sense to leave them aside if they require something too recent.

Thanks!

**Comment by qmonnet** [comment]
Note: If we accept the file, this PR will need a related entry in `CODEOWNERS` (assigning reviews to `@cilium/contributing`, most likely).

File: .clang-format (hunk 7 above)
Comment (bleggett) [reply .clang-format:40]: Bitfields: Yeah looks like we either live with this, or we avoid multiple bitfield declarations in a single statement, e.g. instead of:

```
__u8		deny:1,
		wildcard_protocol:1, /* protocol is fully wildcarded */
		wildcard_dport:1, /* dport is fully wildcarded */
```

we go with 

```
__u8		deny:1;
__u8		wildcard_protocol:1; /* protocol is fully wildcarded */
__u8		wildcard_dport:1; /* dport is fully wildcarded */
```

Either way, clang doesn't have a way to handle both and `checkpatch` doesn't care, so for now it has to be manually reconciled.

Comment (bleggett) [reply .clang-format:40]: Label space:  that seems to be controlled by `SpaceBeforeInheritanceColon` - which seems like a bug, but setting `SpaceBeforeInheritanceColon: false` is probably fine for us given that we don't (afaik) use any class inheritance in bpf code. 

I'm going to set `SpaceBeforeInheritanceColon: false` and add a comment indicating this quirk.

File: .clang-format (hunk 8 above)
Comment (bleggett) [reply .clang-format]: I added some comments around how PenaltyXXX settings work.

I tuned them a _bit_, mostly to get `PenaltyReturnTypeOnItsOwnLine` to stick to existing code style - feels a bit better now but we can twiddle these weights more.

**Comment by bleggett** [comment]
I've bumped the CODEOWNERS as requested, and rolled in most of the changes requested - feel like it's in a good spot and we can tweak further once it's in place if everyone is good with this as a starting point.

**Comment by qmonnet** [comment]
> I've bumped the CODEOWNERS as requested, and rolled in most of the changes requested - feel like it's in a good spot and we can tweak further once it's in place if everyone is good with this as a starting point.

Thanks!

[Regarding `IndentPPDirectives`](https://github.com/cilium/cilium/pull/25488#discussion_r1200256144), I think I'd be in favour of the `AfterHash` value. We could update it later, but I'm thinking we might as well avoid losing the existing indent everywhere we touch the code, so might be worth updating in this PR? Depending on what you think of it, too.

Other than that, agreed, it's in a good shape and we can tweak further later if necessary.

One last thing before we go ahead: Could you please split the commit in two, one with the initial config file copied from the kernel (but without the macro allow-list, so just like in your first version), and then a second one that brings the changes we've agreed on? This way it will be easier to track down the changes we've added for Cilium. I'd also like to have the kernel commit (at which you picked the file) referenced in the description of the commit adding the initial file, so we can track changes on kernel's `.clang-format` if we ever need to.

**Comment by bleggett** [comment]
> [Regarding `IndentPPDirectives`](https://github.com/cilium/cilium/pull/25488#discussion_r1200256144), I think I'd be in favour of the `AfterHash` value. We could update it later, but I'm thinking we might as well avoid losing the existing indent everywhere we touch the code, so might be worth updating in this PR? Depending on what you think of it, too.

👍 

> One last thing before we go ahead: Could you please split the commit in two, one with the initial config file copied from the kernel (but without the macro allow-list, so just like in your first version), and then a second one that brings the changes we've agreed on? This way it will be easier to track down the changes we've added for Cilium. I'd also like to have the kernel commit (at which you picked the file) referenced in the description of the commit adding the initial file, so we can track changes on kernel's `.clang-format` if we ever need to.

Ah good call, that all makes sense, will do.

**Comment by bleggett** [comment]
> > [Regarding `IndentPPDirectives`](https://github.com/cilium/cilium/pull/25488#discussion_r1200256144), I think I'd be in favour of the `AfterHash` value. We could update it later, but I'm thinking we might as well avoid losing the existing indent everywhere we touch the code, so might be worth updating in this PR? Depending on what you think of it, too.
> 
> 👍


Hmm actually trying this on e.g. `bpf_lxc.c` it actually creates some odd-looking results with current code (e.g. line 73-76), so maybe we should leave it off for now?

**Comment by bleggett** [comment]
@qmonnet Commits re-organized as you asked, please let me know about IndentPPDirectives, it seems to mangle existing formatting a bit too much, but if you want it I will add it.

**Review by qmonnet — APPROVED** [review]
Looks all good, thanks a lot!

> please let me know about `IndentPPDirectives: AfterHash`

OK let's leave it out for now, or let's see if others chime in. Hard to tell what option would be the best here anyway.

The checkpatch workflow reports that your commit subject is too long on your first commit, I think it's because you're just missing the empty line between the subject and the description (the link).

Note to other reviewers: This PR was discussed at the community meeting last Wednesday (2023-05-24) and the consensus was to add `.clang-format`, nobody raised any particular objections.

**Comment by bleggett** [comment]
> The checkpatch workflow reports that your commit subject is too long on your first commit, I think it's because you're just missing the empty line between the subject and the description (the link).

Ah,  missed that, fixed, ty!

**Review by julianwiedmann — APPROVED** [review]
